### ISSUE TITLE
Coway IoCare now Rate Limiting

### ISSUE THREAD

**Issue by matthewpucc** [RobertD502/home-assistant-iocare#54]
Recently, I have been getting an error message "Your request has been blocked due to exceeding the API rate limit. (Excessive requests)"

I think the code needs to update the polling rate or implement backoff.

**Comment by RobertD502** [comment]
How many devices do you have on your account? I have 3 and have noticed that this seems to be short-lived and correlated to when they have temporary server stability issues (app is also acting up at the same time).

If you want to manually change the polling frequency, change the DEFAULT_SCAN_INTERVAL variable in `const.py`.

**Comment by gh69** [comment]
I've just noticed this error today as well.  I also have 3 devices with no changes to my setup in the last several months.

**Comment by matthewpucc** [comment]
I only have 2 devices. @RobertD502 have you seen this recently? I've been rate limited the last 18 hours and multiple times over the last week.  I'm giving a go at configurable intervals but won't have much time to dedicate to it till tomorrow or monday.

**Comment by matthewpucc** [comment]
@gh69 report back on if that helps. I was going to give 120s a go since I don't really need _that_ granular of metrics.

**Comment by gh69** [comment]
@matthewpucc I'm not sure how long I'll have to wait to fall off of their rate-limit block-list but hopefully by tomorrow I should know whether or not such a small change made any difference.  If not, I'll just keep bumping it up until it does.  I'll let you know what I find.  @RobertD502 Do you have a recommended alternative timeout value?

**Comment by matthewpucc** [comment]
My plan was to leave it off overnight entirely in hopes of falling off the rate limiting radar. I'll post back if that worked.

**Comment by randomstring** [comment]
I have two devices and have been seeing this issue as well. When I go to use the app, I'm forced to log in again and I get a pop-up saying "This service is restricted due to unauthorized use. Please contact customer service. (Error code:F0101)"

I'm changing the scan interval to 60s and see if that helps.

**Comment by randomstring** [comment]
After a HA restart I saw that the integration startup was failing with the same Error code F0101, and mentions exceeding the API rate limit (unlike the iphone app). The integration setup retry interval is pretty short, maybe < 30s, thus probably perpetuating the API rate limit exception. Possibly indefinitely. Is there some sort of increasing back off interval algorithm into the integration startup? I would hope that HA might have that baked in and not rely on every integration to re-implement it. I temporarily disabled the plugin and will re-enable this afternoon and hope I've cleared the API rate limiting.

Has anyone tried contacting support at Coway? I know, I know, what are the chances you'll end up communicating with someone who knows what you're talking about, but worth a try?

**Comment by matthewpucc** [comment]
I had success with leaving it off overnight and re-enabling. I'm going to set the polling to a 300s interval and see how the day goes.

**Comment by gh69** [comment]
I didn't turn mine off last night...just updated the scan interval from 30 to 45 and left it running and today I'm back to normal with no errors.

**Comment by matthewpucc** [comment]
noice. if you see any issues today, let me know. i may drop the polling rate to 90s or something

**Comment by randomstring** [comment]
The crazy thing is, when I log into the iPhone app it still asks me first if I want to change my password before it kicks me out with the F0101 error. What is wrong with these people?

**Comment by matthewpucc** [comment]
@randomstring , have you tried disabling the integration for a bit?

**Comment by randomstring** [comment]
> @randomstring , have you tried disabling the integration for a bit?

I did. 2 hours later, still blocked. Maybe I can infer how long the API rate limiting is (there could be a couple ways they track API rate). It might be queries per day and clear at midnight (UTC? server's timezone?)

**Comment by randomstring** [comment]
Back online. It took between 2-6 hrs to escape the API rate limit jail.

**Comment by RobertD502** [comment]
It finally hit me as well....except I can't even log into the app. Best thing to do, for now, is to disable the integration and give it a few hours. In the meantime, change the scan interval found in `const.py` to a higher value, save it, and restart Home Assistant. This way, when you re-enable the integration at a later time, the integration will use the new polling interval.

All of this is pretty odd, given that I recall their app has a setting where you can have the app refresh (poll) every 5 seconds.

**Comment by npawelek** [comment]
What's the consensus on polling time here? Would 60s be sufficient or does it need to be higher? Curious what other folks have had success with?

**Comment by randomstring** [comment]
I've been running with the timeout set to 60s without issue for 24 days.  I have two Coway air purifiers.

**Comment by gh69** [comment]
I'd been running with a timeout of 45s (using 3 devices) for about a week before I ran into the issue again so I bumped it up to 300 and have been good ever since.  It's been about 2 weeks now with no issues at 300.

**Comment by RobertD502** [comment]
Wish I could help, but I've been locked out of my account and have been waiting for their support to resolve it for a month now...

**Comment by RobertD502** [comment]
From their support:

```log
There has been an iocare error that was from HQ error which many customers have experienced. They have contacted us telling that they're trying to fix at their very best. In the meantime they also instructed that deleting the app and redownloading might help some users. We sincerely apologize for this matter.
```

In my case, I can't even log into the app (the login form fields reset when trying to log in with no error message popping up).

I'm betting that, aside from some users not being able to log in at all, the rate limiting is also erroneous and part of bugs they introduced in their backend.

**Comment by webdevnerdstuff** [comment]
I have the same problem. Seems like they are in no rush to fix it since it's been an issue for over a month now.

**Comment by papsg** [comment]
Just ran into the same issue. Set the scan interval to 300 and disabled the plugin for now. Hopefully I haven’t been permanently locked out. Is the permanent lockout something you can get around by making another account or are the devices permanently bound to the first account they are registered to?

**Comment by RobertD502** [comment]
I believe you can create another account, but it will require support to unlink devices from the original account - I may actually have to do this because it is coming up on 2 months since I've been able to log in. They state it is a bug on their end that they are trying to fix, but a bug this big should not be taking 2 months to fix.

**Comment by papsg** [comment]
> I believe you can create another account, but it will require support to unlink devices from the original account - I may actually have to do this because it is coming up on 2 months since I've been able to log in. They state it is a bug on their end that they are trying to fix, but a bug this big should not be taking 2 months to fix.

I think I caught it quick enough to not get locked out permanently. I changed the threshold to 600 seconds since I don't have a huge use for the polling data, I just need the buttons in home assistant for workflows. 

I'm a new Coway Airmega user and really appreciate your integration work here which is why I immediately downloaded it. Would you consider changing the default polling threshold in your project so new users don't get caught off guard and locked out. It immediately impacted me a few hours after installation.

**Comment by webdevnerdstuff** [comment]
I was about to create a new account, as that's what their support said to do, but no sure what made me notice, but there was an update for the app. After I updated the app, I was able to log back in again. Not sure if that'll work for everybody, but worth a shot?

**Comment by RobertD502** [comment]
> I was about to create a new account, as that's what their support said to do, but not sure what made me notice, but there was an update for the app. After I updated the app, I was able to log back in again. Not sure if that'll work for everybody, but worth a shot?

No luck on my end on the latest iOS app. What were your symptoms before that? For me, after filling out my credentials, the username and password fields would just clear themselves after I pressed on the Login button. No error message. Nothing.

**Comment by webdevnerdstuff** [comment]
Actually now that I updated HA... the problem is back. Seems kind of dumb to have to create another account.

**Comment by matthewhadley** [comment]
Adding another data point as I just hit this API rate limit - seeing the "The service is restricted due to unauthorized use. Please contact customer service. (error code F0101)".

I've been running with one 205S device, but bought another one yesterday and got the error today. I've disabled the integration for now, bumped up the `DEFAULT_SCAN_INTERVAL` (to 60) and will re-enable tomorrow and see how things are.

**Comment by matthewhadley** [comment]
Things are back working today with the the `60` value, but I agree with @papsg and have now set it to `600` as I only interact with the devices via home assistant and not the native app.

**Comment by RobertD502** [comment]
> > > I was about to create a new account, as that's what their support said to do, but not sure what made me notice, but there was an update for the app. After I updated the app, I was able to log back in again. Not sure if that'll work for everybody, but worth a shot?
> 
> > 
> 
> > No luck on my end on the latest iOS app. What were your symptoms before that? For me, after filling out my credentials, the username and password fields would just clear themselves after I pressed on the Login button. No error message. Nothing.
> 
> > 
> 
> > Edit: Disregard. Grabbed an android device and used the latest app - same issue. So, my account issue is part of a broader bug related to their backend and not something to do with the app. Hopefully one of these days support will get back to me.
> 
> 
> 
> I've been unable to login for awhile now too and finally getting around to checking on this (busy summer). I can reset my password but I encounter the same symptoms when using the iOS app: the username/password "clear" and the HA app says 'unable to retrieve code'.
> 
> 
> 
> What POS devices! I have 2 of these 400s. I'm giving serious thought to crushing them with my tractor, then throw on some gasoline and light them on fire--- other ideas? Maybe if I do something crazy enough and gain some traction on social media they'll have to take notice and then you guys will get fixed 🤣

They finally responded to me with the suggestion of RMAing my units....which doesn't solve not being able to log in. Hopefully I can get them to finally unlink my purifiers so that I can add them to a new account. Feels like pulling teeth asking for a simple solution to be carried out.

**Comment by RobertD502** [comment]
@cw-cdz 

Support unlinked my purifiers and I was able to create a new account + add the purifiers to it.
I know it wasn't the case before, but it looks like it may be possible to unlink the purifiers from your old account without contacting support. 

When adding each purifier to the new account, I was greeted with this option which seems to indicate that you can unlink the purifiers without needing to play telephone with support:

![image](https://github.com/user-attachments/assets/b06ee39e-22e5-4cb6-abce-522095859611)

**Comment by cmaglio** [comment]
Hey guys, I just want to add my experience here with this issue and how I fixed it. A month or two ago my Coway account was also locked. I assume this was because of the integration/rate limit issue. I haven't been able to get into that account for a couple of months now.

I finally got fed up with not being able to control the devices, and considering my previous support experiences with Coway have been less than stellar, I decided to just create a new account and try registering my products to that account. Luckily, I was able to register a new account by using a different email address (just a different variation - added a dot - to my already registered gmail address). I was then able to scan the barcode of each product and the app asked me if I wanted to delete the previous user from the device. I was able to successfully add all of my Coway devices to this new account and then re-authenticate the HA integration.

Long story short, I was able to fix this myself by registering a new account under a different email address and re-adding devices without any intervention by Coway Support. Honestly, it's probably the easiest and fastest way to fix this given some previous horrible and drawn out experiences with Coway's support team.

**Comment by RobertD502** [comment]
> Hey guys, I just want to add my experience here with this issue and how I fixed it. A month or two ago my Coway account was also locked. I assume this was because of the integration/rate limit issue. I haven't been able to get into that account for a couple of months now.
> 
> 
> 
> I finally got fed up with not being able to control the devices, and considering my previous support experiences with Coway have been less than stellar, I decided to just create a new account and try registering my products to that account. Luckily, I was able to register a new account by using a different email address (just a different variation - added a dot - to my already registered gmail address). I was then able to scan the barcode of each product and the app asked me if I wanted to delete the previous user from the device. I was able to successfully add all of my Coway devices to this new account and then re-authenticate the HA integration.
> 
> 
> 
> Long story short, I was able to fix this myself by registering a new account under a different email address and re-adding devices without any intervention by Coway Support. Honestly, it's probably the easiest and fastest way to fix this given some previous horrible and drawn out experiences with Coway's support team.

That is exactly what I did as well. Just to add, the account not working isn't related to the integration or rate limiting. The coway devs messed up something on their end that caused the issue for some and were actively trying to fix the issue - I got fed up with waiting and used a new account. This isn't the first time a backend bug was introduced by their dev team. Last time they did some major work, one of the purifiers on my account was missing for weeks until they tracked down the bug in their code.

I've been testing a 90 second interval for a bit now with no issues. I'll be releasing that as the default soon and also adding an option for it in the integration UI for those who wish to change it to something else.

**Comment by cmaglio** [comment]
Got it, thanks for the clarification. At one point the app was giving me a message that gave the impression the account was locked due to suspicious or otherwise irregular activity. I don't remember what the exact wording was, but I had the perception it was something I did. Good to know it wasn't! Thanks again for all your work on this integration Robert, it's been great!

**Comment by cmaglio** [comment]
@RobertD502 Are you sure they aren't mistaking the integration for some type of unauthorized use? My new account just got locked tonight only 3 hours after reauthorizing the integration. Here is the error I'm seeing in the IOCare app:

![image](https://github.com/user-attachments/assets/0fd553d2-a1f1-400d-ba1b-6b773e8a3729)

**Comment by RobertD502** [comment]
> @RobertD502 Are you sure they aren't mistaking the integration for some type of unauthorized use? My new account just got locked tonight only 3 hours after reauthorizing the integration. Here is the error I'm seeing in the IOCare app:
> 
> ![image](https://github.com/user-attachments/assets/0fd553d2-a1f1-400d-ba1b-6b773e8a3729)

Did you change the polling interval to a higher value as discussed earlier in this issue? I'd disable the integration in the meantime and then, after 24 hours, try to log into your account using the app (not the integration). If it purely rate limiting then the rate limiting will fall off and you'll be able to log into again.

The account not working at all had different symptoms - no error message being disabled, just the app resetting the email and password fields when pressing the login button.

**Comment by cmaglio** [comment]
Above is the same message I saw months ago when this issue first started. But, I have also experienced the login fields resetting you describe when trying to login on that old account. I wonder if the two symptoms are linked but don't always reproduce together.

I did not change the polling rate tonight (I must've missed that functionality was already released). All I did was reauthorize with the new account.

I disabled the integration and will wait 24 hours to see if it clears. If not, then I'll create a third account and slow down the polling to retest if that one gets locked.

Thanks for the advice!

**Comment by RobertD502** [comment]
For now, you'll need to change the polling frequency in the cost.py file. I will release a higher default polling interval and being able to change the polling interval via the UI in the future.

**Comment by cmaglio** [comment]
I was able to access the new account again this morning, so it seems the default polling rate must be causing an issue. I am not sure if it matters, but I have 3 devices on the account (400S Air Purifiers)... so I'm not sure if having more devices plus the higher polling rate is compounding the problem. I'll go in today to change the polling rate to 90 seconds, re-enable the integration, and report back.

**Comment by cmaglio** [comment]
So far so good at the 90 second scan interval today, may increase it to 120 to allow for some headroom since others have had some issues at 90/95.

**Comment by RobertD502** [comment]
> So far so good at the 90 second scan interval today, may increase it to 120 to allow for some headroom since others have had some issues at 90/95.

I misspoke and have been testing at 120 seconds. The version just released defaults to 120. However, you can set a different polling interval, if you wish, via the options UI.
